wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/DefaultEditorProvisioner.java

```diff
@@ -0,0 +1,127 @@
+      @Named("che.workspace.devfile.default_editor") String defaultEditorRef,
+      @Named("che.workspace.devfile.default_editor.plugins") String[] defaultPluginsRefs) {
+    this.defaultEditorRef = Strings.isNullOrEmpty(defaultEditorRef) ? null : defaultEditorRef;
+    this.defaultEditorId = this.defaultEditorRef == null ? null : getId(this.defaultEditorRef);
+    this.defaultPluginsIdToRef =
+        Arrays.stream(defaultPluginsRefs).collect(toMap(this::getId, identity()));
+  }
+
+  /**
+   * Provision default editor if there is no any another editor and default plugins for it.
+   *
+   * @param devfile devfile where editor and plugins should be provisioned
+   */
+  public void apply(Devfile devfile) {
+    if (defaultEditorRef == null) {
+      // there is no default editor configured
+      return;
+    }
+
+    List<Component> components = devfile.getComponents();
+    Set<String> componentsNames =
+        components.stream().map(Component::getName).collect(Collectors.toCollection(HashSet::new));
+
+    Optional<Component> editorOpt =
+        components.stream().filter(t -> EDITOR_COMPONENT_TYPE.equals(t.getType())).findFirst();
+
+    boolean isDefaultEditorUsed;
+    if (!editorOpt.isPresent()) {
+      components.add(
+          new Component()
+              .withName(findAvailableName(componentsNames, defaultEditorRef))
+              .withType(EDITOR_COMPONENT_TYPE)
+              .withId(defaultEditorRef));
+      isDefaultEditorUsed = true;
+    } else {
+      Component editor = editorOpt.get();
+      isDefaultEditorUsed = editor.getId().startsWith(defaultEditorId + ':');
+    }
+
+    if (isDefaultEditorUsed) {
+      provisionDefaultPlugins(components, componentsNames);
+    }
+  }
+
+  private void provisionDefaultPlugins(List<Component> components, Set<String> componentsNames) {
+    Map<String, String> missingPluginsIdToRef = new HashMap<>(defaultPluginsIdToRef);
+
+    components
+        .stream()
+        .filter(t -> PLUGIN_COMPONENT_TYPE.equals(t.getType()))
+        .forEach(t -> missingPluginsIdToRef.remove(getId(t.getId())));
+
+    missingPluginsIdToRef
+        .values()
+        .forEach(
+            pluginRef ->
+                components.add(
+                    new Component()
+                        .withType(PLUGIN_COMPONENT_TYPE)
+                        .withId(pluginRef)
+                        .withName(findAvailableName(componentsNames, pluginRef))));
+  }
+
+  /**
+   * Returns available name for component with the specified id.
+   *
+   * <p>Id without version is used as base name and generated part is added if it is already busy.
+   */
+  private String findAvailableName(Set<String> busyNames, String componentRef) {
+    String id = getId(componentRef);
+    String name = id;
+    while (!busyNames.add(name)) {
+      name = NameGenerator.generate(id, 5);
+    }
+    return name;
+  }
+
+  private String getId(String reference) {
+    return reference.split(":", 2)[0];
```

Comment:
Are we assuming the version can contain colons?

Reply:
It must not 
https://github.com/eclipse/che/blob/26611a5762af998e09be9c1882a9ba79b7deda96/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsplugins/PluginMetaRetriever.java#L156-L161

I think it makes sense to add field pattern to JsonSchema for such check. @metlos WDYT?

Reply:
That check has not yet been performed at the time we're applying the default editor. Therefore, I think adding the pattern to the schema is a good idea.